File: packages/llm/src/summarize.ts
import { type Result, ok, err } from "@repo/shared";
import { complete } from "./client.js";

export interface SummaryInput {
  productName: string;
  productUrl: string;
  pageTitle: string;
  headings: string[];
  bodyText: string;
  loadTimeMs: number;
}

export interface SummaryResult {
  content: string;
  targetAudience: string;
  keyFeatures: string[];
  pros: string[];
  cons: string[];
  model: string;
}

function parseJsonSafe(raw: string): Record<string, unknown> | null {
  try {
    return JSON.parse(raw);
  } catch {
    // Fallback: try extracting JSON from code blocks
    const match = raw.match(/\{[\s\S]*\}/);
    if (match) {
      try {
        return JSON.parse(match[0]);
      } catch {
        return null;
      }
    }
    return null;
  }
}

export async function summarizeProduct(input: SummaryInput): Promise<Result<SummaryResult>> {
  const model = "claude-sonnet-4-5-20250929";
  const truncatedBody = input.bodyText.slice(0, 8000);

  const prompt = `You are a B2B software reviewer. Analyze this product and provide a structured review.

Product: ${input.productName}
URL: ${input.productUrl}
Page Title: ${input.pageTitle}
Headings: ${input.headings.join(", ")}
Page Load Time: ${input.loadTimeMs}ms

Page Content (truncated):
${truncatedBody}

Respond in this exact JSON format (no markdown, just JSON):
{
  "content": "A 2-3 paragraph review of the product",
  "targetAudience": "Who this product is best for",
  "keyFeatures": ["feature1", "feature2"],
  "pros": ["pro1", "pro2"],
  "cons": ["con1", "con2"]
}`;

  const rawResult = await complete(prompt, { model });
  if (!rawResult.ok) {
    return rawResult;
  }

  const parsed = parseJsonSafe(rawResult.value);

  if (!parsed) {
    return err("Failed to parse LLM summary response as JSON");
  }

  return ok({
    content: String(parsed.content ?? ""),
    targetAudience: String(parsed.targetAudience ?? ""),
    keyFeatures: Array.isArray(parsed.keyFeatures) ? parsed.keyFeatures.map(String) : [],
    pros: Array.isArray(parsed.pros) ? parsed.pros.map(String) : [],
    cons: Array.isArray(parsed.cons) ? parsed.cons.map(String) : [],
    model,
  });
}

File: apps/go-api/internal/llm/summarize.go
package llm

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

const maxBodyTextChars = 8000

// SummaryInput contains the page data needed to generate a product summary.
type SummaryInput struct {
	ProductName string
	ProductURL  string
	PageTitle   string
	Headings    []string
	BodyText    string
	LoadTimeMs  int
}

// SummaryResult holds the structured output of a product summary.
type SummaryResult struct {
	Content        string   `json:"content"`
	TargetAudience string   `json:"targetAudience"`
	KeyFeatures    []string `json:"keyFeatures"`
	Pros           []string `json:"pros"`
	Cons           []string `json:"cons"`
	Model          string   `json:"model"`
}

// Summarize generates an AI product summary from page extraction data.
// Truncates body text to 8000 chars to stay within context limits.
func (c *Client) Summarize(ctx context.Context, input SummaryInput) (SummaryResult, error) {
	bodyText := input.BodyText
	if len(bodyText) > maxBodyTextChars {
		bodyText = bodyText[:maxBodyTextChars]
	}

	prompt := fmt.Sprintf(`You are a professional B2B software reviewer. Analyze this product and provide a structured review.

Product: %s
URL: %s
Page Title: %s
Headings: %s
Body Text: %s
Load Time: %dms

Respond with ONLY valid JSON in this exact format:
{
  "content": "A 2-3 paragraph professional review of the product",
  "targetAudience": "Who this product is best suited for",
  "keyFeatures": ["feature1", "feature2", "feature3"],
  "pros": ["pro1", "pro2", "pro3"],
  "cons": ["con1", "con2", "con3"]
}`,
		input.ProductName,
		input.ProductURL,
		input.PageTitle,
		strings.Join(input.Headings, ", "),
		bodyText,
		input.LoadTimeMs,
	)

	raw, err := WithRetry(ctx, func() (string, error) {
		return c.Complete(ctx, prompt, WithSystem("You are a B2B SaaS product reviewer. Always respond with valid JSON only."))
	}, DefaultRetryOptions())
	if err != nil {
		return SummaryResult{}, fmt.Errorf("summarize completion: %w", err)
	}

	var result SummaryResult
	if err := json.Unmarshal([]byte(raw), &result); err != nil {
		// Fallback: attempt regex extraction for each field.
		result = extractSummaryFallback(raw)
		if result.Content == "" {
			return SummaryResult{}, fmt.Errorf("parsing summary JSON: %w (raw: %.200s)", err, raw)
		}
	}

	result.Model = "claude-sonnet-4-20250514"
	return result, nil
}

// extractSummaryFallback uses regex to pull fields from malformed JSON responses.
func extractSummaryFallback(raw string) SummaryResult {
	return SummaryResult{
		Content:        extractJSONString(raw, "content"),
		TargetAudience: extractJSONString(raw, "targetAudience"),
		KeyFeatures:    extractJSONArray(raw, "keyFeatures"),
		Pros:           extractJSONArray(raw, "pros"),
		Cons:           extractJSONArray(raw, "cons"),
	}
}

func extractJSONString(raw, key string) string {
	re := regexp.MustCompile(fmt.Sprintf(`"%s"\s*:\s*"((?:[^"\\]|\\.)*)"`, key))
	matches := re.FindStringSubmatch(raw)
	if len(matches) > 1 {
		return matches[1]
	}
	return ""
}

func extractJSONArray(raw, key string) []string {
	re := regexp.MustCompile(fmt.Sprintf(`"%s"\s*:\s*\[([^\]]*)\]`, key))
	matches := re.FindStringSubmatch(raw)
	if len(matches) < 2 {
		return nil
	}

	var items []string
	itemRe := regexp.MustCompile(`"((?:[^"\\]|\\.)*)"`)
	for _, m := range itemRe.FindAllStringSubmatch(matches[1], -1) {
		if len(m) > 1 {
			items = append(items, m[1])
		}
	}
	return items
}
